src/Database/LSMTree/Internal/Serialise.hs

```diff
@@ -1,205 +1,160 @@
-     in case compareByteArrays ba1 off1 ba2 off2 len of
-          EQ | len1 < len2 -> LT
-             | len1 > len2 -> GT
-          o  -> o
-  where
-    P.Vector off1 _size1 ba1 = vec1
-    P.Vector off2 _size2 ba2 = vec2
-
-instance Hashable SerialisedKey where
-  hashIO32 :: SerialisedKey -> Word32 -> IO Word32
-  hashIO32 = hashBytes
-
--- TODO: optimisation
-hashBytes :: SerialisedKey -> Word32 -> IO Word32
-hashBytes (SerialisedKey vec) = hashList32 (P.toList vec)
-
--- | @'topBits16' n k@ slices the first @n@ bits from the /top/ of the
--- serialised key @k@. Returns the string of bits as a 'Word16'.
---
--- The /top/ corresponds to the most significant bit (big-endian).
---
--- PRECONDITION: @n >= 0 && n <= 16. We can slice out at most 16 bits,
--- all bits beyond that are truncated.
---
--- PRECONDITION: The byte-size of the serialised key should be at least 2 bytes.
---
--- TODO: optimisation ideas: use unsafe shift/byteswap primops, look at GHC
--- core, find other opportunities for using primops.
---
-topBits16 :: Int -> SerialisedKey -> Word16
-topBits16 n k@(SerialisedKey (P.Vector (I# off#) _size (ByteArray k#))) =
-    assert (sizeofKey k >= 2) $ shiftR w16 (16 - n)
-  where
-    w16 = toWord16 (indexWord8ArrayAsWord16# k# off#)
-
-toWord16 :: Word16# -> Word16
-#if WORDS_BIGENDIAN
-toWord16 = W16#
-#else
-toWord16 x# = byteSwap16 (W16# x#)
-#endif
-
--- | @'sliceBits32' off k@ slices from the serialised key @k@ a string of @32@
--- bits, starting at the @0@-based offset @off@. Returns the string of bits as a
--- 'Word32'.
---
--- Offsets are counted in bits from the /top/: offset @0@ corresponds to the
--- most significant bit (big-endian).
---
--- PRECONDITION: the byte-size @n@ of the serialised key should be at least 4
--- bytes.
---
--- PRECONDITION: The serialised key should be large enough that we can slice out
--- 4 bytes after of the bit-offset @off, since we can only slice out bits that
--- are within the bounds of the byte array.
---
--- TODO: optimisation ideas: use unsafe shift/byteswap primops, look at GHC
--- core, find other opportunities for using primops.
---
-sliceBits32 :: Int -> SerialisedKey -> Word32
-sliceBits32 off@(I# off1#) k@(SerialisedKey (P.Vector (I# off2#) _size (ByteArray k#)))
-    | 0# <- r#
-    = assert (off + 32 <= 8 * sizeofKey k) $
-      toWord32 (indexWord8ArrayAsWord32# k# q#)
-    | otherwise
-    = assert (off + 32 <= 8 * sizeofKey k) $
-        toWord32 (indexWord8ArrayAsWord32# k# q#       ) `shiftL` r
-      + w8w32#   (indexWord8Array#         k# (q# +# 4#)) `shiftR` (8 - r)
-  where
-    !(# q0#, r# #) = quotRemInt# off1# 8#
-    !q#            = q0# +# off2#
-    r              = I# r#
-    -- No need for byteswapping here
-    w8w32# x#     = W32# (wordToWord32# (word8ToWord# x#))
-
-toWord32 :: Word32# -> Word32
-#if WORDS_BIGENDIAN
-toWord32 = W32#
-#else
-toWord32 x# = byteSwap32 (W32# x#)
-#endif
+newtype SerialisedKey = SerialisedKey RawBytes
```

Comment:
Maybe a short comment that specific types of index can have additional requirements, such as a minimum length?